perceval/utils/density_matrix_utils.py

```diff
@@ -0,0 +1,95 @@
+# MIT License
+#
+# Copyright (c) 2022 Quandela
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# As a special exception, the copyright holders of exqalibur library give you
+# permission to combine exqalibur with code included in the standard release of
+# Perceval under the MIT license (or modified versions of such code). You may
+# copy and distribute such a combined system following the terms of the MIT
+# license for both exqalibur and Perceval. This exception for the usage of
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from perceval.utils.statevector import StateVector
+from typing import Union
+import numpy as np
+from scipy.sparse import sparray, csr_array
+
+
+def extract_upper_triangle(csr_matrix: csr_array) -> csr_array:
+    """
+    extract the upper half matrix S of a hermitian matrix rho such that S + S_dagger = rho
```

Comment:
Why does the csr_matrix have to be hermitian? I think in this docstring you're only describing *your* use case of this function. But this function is just extracting the top right half of any square sparse matrix, isn't it?
If so the docstring should just say that and nothing more. 

If you want to tell other devs that you're extracting the triangle of an hermitian matrix because it simplifies the computation, given the S + Sdagger = rho rule, please do so, but not here (I guess that would be in the `Simulator.evolve_density_matrix` method)